File: material/algorithm/leetcode/0006_ConvertZ.java
package material.algorithm.leetcode;

class ConvertZ {
    public static void main(String[] args) throws InterruptedException {
      String s = "LEETCODEISHIRINGIXZXSA";
      String r = convert(s, 5);
      System.out.println(r);
    }
    public static String convert(String s, int numRows) {
      if (numRows == 1) return s;
      // 速度快慢为：StringBuilder > StringBuffer > String
      // String为字符串常量，而StringBuilder和StringBuffer均为字符串变量，
      // 即String对象一旦创建之后该对象是不可更改的，但后两者的对象是变量，是可以更改的。
      // 在线程安全上，StringBuilder是线程不安全的，而StringBuffer是线程安全的
      // String：适用于少量的字符串操作的情况
      // StringBuilder：适用于单线程下在字符缓冲区进行大量操作的情况
      // StringBuffer：适用多线程下在字符缓冲区进行大量操作的情况
      String str = "";
      int len = s.length();
      int cLen = numRows * 2 - 2;
      for (int i = 0; i < numRows; i++) {
        for (int j = 0; j + i < len; j += cLen) {
          str += s.charAt(i + j);
          if (i != 0 && i != numRows - 1 && j + cLen - i < len) {
            str += s.charAt(j + cLen - i);
          }
        }
      }
      return str;
    }
}

File: material/algorithm/leetcode/0006_ConvertZ.js
/**
 * 
 * 
4
 +6
 (0,0)                 (1,6)                 (2,12)

 (3,1)        (4,5)    (5,7)          (6,11) (7,13)            (8,17)

 (9,2)  (10,4)         (11,8)  (12,10)       (13, 14)  (14, 16)
 
 (15,3)                (16,9)                (17, 15)
5
      col1                  col2                   col3
 row1 0,0                  1,8                     2,16
 
 row2 3,1            4 ,7  5,9               6,15, 7,17
 
 row3 8,2        9,6       10,10       14,11       12,18         13,22
 
 row4 14,3  15,5           16,11  17,13            18,19   19,21
 
 row5 20,4                 21,12                   22,20

 * @param {*} s 
 * @param {*} numRows 
 */
var convert = function (s, n) {
  let len = s.length
  let str = ''
  let cLen = (n - 1) * 2
  for (let i = 0; i < n; i++) {
    for (let j = 0; i + j < len; j+=cLen ) {
      str += s[i + j]
      if (i !== 0 && i !== n - 1 && j + cLen - i < len) { // 不是第一行和最后一行
        str += s[j + cLen - 1]
      }
    }
  }
  return str
};

console.log(convert('LEETCODEISHIRINGIXZXSA', 5));